File: shared/desktop/appearance.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Le bandeau de la fenêtre suit le thème du CRM.
 *
 * Le cadre, les feux de circulation et la barre de titre sont dessinés par le
 * système, qui les accorde à l'apparence du **poste**. Un Mac en sombre posait
 * donc un bandeau noir au-dessus d'une application en thème clair : deux
 * moitiés de fenêtre qui ne se ressemblent pas. La coque le règle en une
 * ligne (`src-tauri/src/appearance.rs`) ; encore faut-il lui dire quoi.
 *
 * **On écoute la classe `dark` du document, pas les préférences.** C'est elle
 * le thème *résolu* — « système » y est déjà tranché — et c'est la seule chose
 * que les deux chemins qui l'appliquent ont en commun : `applyPreferences` au
 * rendu, et le script anti-scintillement qui la pose avant la première
 * peinture. Lire les préférences obligerait `shared/desktop` à dépendre d'un
 * module, et manquerait le second chemin.
 */

let last: boolean | null = null;

function push(dark: boolean) {
  // Le système ne fait rien d'un ordre qu'il a déjà reçu, et une bascule de
  // palette change la classe sans changer le thème.
  if (dark === last) return;
  last = dark;
  // Hors de l'application — `bun run dev` dans un navigateur ordinaire — il n'y
  // a aucune coque pour répondre, et ce n'est pas une panne.
  void invoke("window_theme", { dark }).catch(() => {});
}

/**
 * Aligne la fenêtre maintenant, puis à chaque changement de thème. Rend la
 * fonction qui arrête d'écouter.
 */
export function followDocumentTheme(): () => void {
  const root = document.documentElement;
  const sync = () => push(root.classList.contains("dark"));

  sync();
  const observer = new MutationObserver(sync);
  observer.observe(root, { attributes: true, attributeFilter: ["class"] });
  return () => observer.disconnect();
}

File: src-tauri/src/appearance.rs
//! L'apparence de la fenêtre, alignée sur le thème du CRM.
//!
//! Le bandeau de titre, les feux de circulation et le cadre sont dessinés par
//! le système, pas par la page : ils suivent l'apparence du poste. Quelqu'un
//! dont le Mac est en sombre et qui travaille en thème clair voyait donc un
//! bandeau noir posé sur une application blanche — deux moitiés de fenêtre qui
//! ne se ressemblent pas.
//!
//! `set_theme` dit au système quelle apparence donner à **cette** fenêtre. Ce
//! n'est pas le thème de la page : c'est ce qui entoure la page, et il n'y a
//! rien d'autre à changer pour que les deux s'accordent.
//!
//! Le choix vit dans le stockage local du navigateur — la coque ne peut pas le
//! lire — donc c'est la page qui l'annonce, et elle le fait dès son premier
//! rendu.

use tauri::{AppHandle, Manager, Runtime, Theme};

/// Aligne la fenêtre sur le thème que la page vient d'appliquer.
///
/// Sur macOS et Linux, l'apparence vaut pour toute l'application et non pour
/// la seule fenêtre — sans conséquence ici, l'application n'en a qu'une.
pub fn apply<R: Runtime>(app: &AppHandle<R>, dark: bool) -> Result<(), String> {
    let window = app
        .get_webview_window("main")
        .ok_or_else(|| "fenêtre « main » introuvable".to_string())?;

    let theme = if dark { Theme::Dark } else { Theme::Light };
    window
        .set_theme(Some(theme))
        .map_err(|e| format!("apparence de la fenêtre : {e}"))
}
